api/client.go

```diff
@@ -654,7 +659,28 @@ func websocketDialWithErrors(dialer WebsocketDialer, urlStr string, requestHeade
 		}
 		return nil, err
 	}
-	return c, nil
+	result := DeadlineStream{Conn: c, Timeout: websocketTimeout}
+	return &result, nil
+}
+
+// DeadlineStream wraps a websocket connection and applies a write
+// deadline to each WriteJSON call.
+type DeadlineStream struct {
+	*websocket.Conn
+
+	Timeout time.Duration
+}
+
+// WriteJSON is part of base.Stream.
+func (s *DeadlineStream) WriteJSON(v interface{}) error {
+	// This uses a real clock rather than trying to use a clock passed
+	// in because the websocket will use a real clock to determine
+	// whether the deadline has passed anyway.
+	deadline := time.Now().Add(s.Timeout)
+	if err := s.Conn.SetWriteDeadline(deadline); err != nil {
```

Comment:
this isn't threadsafe - can it be used by more than one thread?

Reply:
It can't be, no. I've gone through the uses of ConnectStream/ConnectControllerStream to see whether it is, these are the uses I can see:

* pubsub worker - only writes from one goroutine
* log forwarder worker - sends to syslog, only reads from the stream
* migrationmaster worker - only writes from one goroutine
* debug-log command - doesn't use WriteJSON, only reads
* logsender - only writes from one goroutine

Which obviously doesn't guard against someone writing to one from multiple goroutines in the future. I'll add a comment to the docstrings of Connect*Stream that writing should only be done from one goroutine because of the deadline.